### ISSUE TITLE
Sumologic fieldmappings and custom parsing

### ISSUE THREAD

**Issue by juju4** [Neo23x0/sigma#235]
One problem that I didn't solve for now to improve sumologic support is to implement fieldmappings with custom parsing.
As said, sumologic support some field extraction rule (FER) at log ingestion but there is a limit and while it is enough to address top sigma fields, it won't be for the numerous windows fields or other applications. But it supports inline parsing. That's what I am looking to implement if possible with a bit of help :)

example on EventID=4624
(native sumo windows collection)
```
_index=WINDOWS EventID=4624
| parse "\tInsertionStrings = {\"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\", \"*\"};" as SubjectUserSid,SubjectUserName,SubjectDomainName,SubjectLogonId,TargetUserSid,TargetUserName,TargetDomainName,TargetLogonId,LogonType,LogonProcessName,AuthenticationPackageName,WorkstationName,LogonGuid,TransmittedServices,LmPackageName,KeyLength,ProcessId,ProcessName,IpAddress,IpPort,ImpersonationLevel,RestrictedAdminMode,TargetOutboundUserName,TargetOutboundDomainName,VirtualAccount,TargetLinkedLogonId,ElevatedToken
```
(through nxlog json collection )
```
_index=NXLOG_WINDOWS EventID=4624
| json "SourceName","EventID","SubjectUserName", "SubjectDomainName", "SubjectLogonId", "TargetUserSid", "TargetUserName", "TargetDomainName", "TargetLogonId","LogonType","LogonProcessName","AuthenticationPackageName","WorkstationName","LogonGuid","TransmittedServices","LmPackageName","KeyLength","ProcessID","ProcessName","IpAddress","IpPort","ImpersonationLevel","RestrictedAdminMode","TargetOutboundUserName","TargetOutboundDomainName","VirtualAccount","TargetLinkedLogonId","ElevatedToken"
```

after a parse statement, matches would become from
```
field1=string1 field2=str*ing2
```
to
```
| where field1="string1" and field2 matches "str*ing2"
```
Note that full text search should be done before parsing or matched against integrated field _raw

So inside fieldmappings backend config, I want to do something like following as example
```
fieldmappings
  EventID
    - 4624: _index=NXLOG_WINDOWS EventID=4624
| json "SourceName","EventID","SubjectUserName", "SubjectDomainName", "SubjectLogonId", "TargetUserSid", "TargetUserName", "TargetDomainName", "TargetLogonId","LogonType","LogonProcessName","AuthenticationPackageName","WorkstationName","LogonGuid","TransmittedServices","LmPackageName","KeyLength","ProcessID","ProcessName","IpAddress","IpPort","ImpersonationLevel","RestrictedAdminMode","TargetOutboundUserName","TargetOutboundDomainName","VirtualAccount","TargetLinkedLogonId","ElevatedToken"
```

**Comment by thomaspatzke** [comment]
Discussed some similar requirement few months ago with a Sigma user. Idea was to add something to the converter configuration that allows to specify a static prefix that is put before the query. The initial idea was to bind these query prefixes to specific log sources. If I understand your requirement correctly, this would be solved by another possibility to bind generated query prefixes to matched event attributes?

E.g. something like:
```
query_prefixes:
    EventID=4623: '| json "SourceName","EventID","SubjectUserName", "SubjectDomainName", "SubjectLogonId", "TargetUserSid", "TargetUserName", "TargetDomainName", "TargetLogonId","LogonType","LogonProcessName","AuthenticationPackageName","WorkstationName","LogonGuid","TransmittedServices","LmPackageName","KeyLength","ProcessID","ProcessName","IpAddress","IpPort","ImpersonationLevel","RestrictedAdminMode","TargetOutboundUserName","TargetOutboundDomainName","VirtualAccount","TargetLinkedLogonId","ElevatedToken"'
```

After conversion, this would prepend the generated query.

**Comment by juju4** [comment]
Yes. I do a suffix part per rule file inside my contrib test script (https://github.com/Neo23x0/sigma/pull/234/commits/98a18fd4a28db2423481ea44a2d04fc14c8b71d6#diff-22067229bf679a92138bf5d2a120daf8R168). Mostly for contextual whitelisting and formatting.

but they are part of sumologic query which must be either at the beginning or "middle".

your example is good but query format/optimization makes it a little bit more complex as you can't just put matching fields after.

Example
```
EventID=4623 string1 string2 field1=*string3*
```
would be
```
EventID=4623 string1 string2 string3
| json "field1", "field2"
| where field1 matches "*string3*"
```
first-line full-text string match is for optimization (and is case-insensitive for sumologic)
after parsing element (json or parse operator), all matches must be as where statement, either `field1="exactvalue"`, either `field2 matches "*wildcard*text*"`. currently no match regex operator (only parse regex operator but it's another story)

```
EventID=4623
| json "field1", "field2"
| string1 string2 field1=*string3*
```
=> would be invalid

On windows, parsing statement is mostly tied to EventID and ingestion format like json
It can also apply to osquery json results depending on pack or auditd keyvalue.

**Comment by thomaspatzke** [comment]
I think this is too specific to be covered by a generic prefixing or templating, as the query condition appears in two different forms. I think the best way would be to integrate it directly into the SumoLogic backend.

**Comment by juju4** [comment]
ok. any advices on where it is best implemented? existing function or similar code? or best to avoid breaking everything ;-)

like fieldMappings in wdatp.py?
